backend/arm64/isa/arm64_ast.ml

```diff
@@ -0,0 +1,597 @@
+    let ppf = Format.formatter_of_out_channel oc in
+    print_program ppf lines
+end
+
+module DSL = struct
+  (* Statically allocate some common combinations *)
+  let reg_array size name = Array.init size (fun i -> Reg.create name i)
+
+  let reg_and_operand_array ~last name =
+    let size = last + 1 in
+    let reg_array = reg_array size name in
+    let op_array = Array.init size (fun i -> Operand.Reg reg_array.(i)) in
+    reg_array, op_array
+
+  let operand_array ~last name =
+    let _reg_array, op_array = reg_and_operand_array ~last name in
+    op_array
+
+  let neon_operand_array name =
+    operand_array ~last:Neon_reg_name.last (Reg_name.Neon name)
+
+  let gp_reg_and_operand_array name =
+    reg_and_operand_array ~last:GP_reg_name.last_numbered (Reg_name.GP name)
+
+  let reg_x, reg_x_operands = gp_reg_and_operand_array GP_reg_name.X
+
+  let _, reg_w_operands = gp_reg_and_operand_array GP_reg_name.W
+
+  let reg_v2d_operands = neon_operand_array Neon_reg_name.(Vector V2D)
+
+  let reg_s_operands = neon_operand_array Neon_reg_name.(Scalar S)
+
+  let reg_d_operands = neon_operand_array Neon_reg_name.(Scalar D)
+
+  let reg_q_operands = neon_operand_array Neon_reg_name.(Scalar Q)
+
+  let mem ~base ~offset =
+    Operand.(Mem (Addressing_mode.Offset (reg_x.(base), offset)))
+
+  let mem_pre ~base ~offset =
+    Operand.(Mem (Addressing_mode.Pre (reg_x.(base), offset)))
+
+  let mem_post ~base ~offset =
+    Operand.(Mem (Addressing_mode.Post (reg_x.(base), offset)))
+
+  let literal l = Operand.(Mem (Addressing_mode.Literal l))
+
+  let reg_v2s index =
+    Operand.Reg (Reg.create (Reg_name.Neon (Vector V2S)) index)
+
+  let reg_v4s index =
+    Operand.Reg (Reg.create (Reg_name.Neon (Vector V4S)) index)
+
+  let reg_v2d index = reg_v2d_operands.(index)
+
+  let reg_s index = reg_s_operands.(index)
+
+  let reg_d index = reg_d_operands.(index)
+
+  let reg_q index = reg_q_operands.(index)
+
+  let reg_w index = reg_w_operands.(index)
+
+  let reg_x index = reg_x_operands.(index)
+
+  let reg_w index = reg_w_operands.(index)
+
+  let imm n = Operand.Imm n
+
+  let ins name operands = Asm.Ins { name; operands }
+
+  let ins_cond name cond operands =
+    let operands = Array.append operands [| Operand.Cond cond |] in
+    ins name operands
+
+  let print_line line = Format.asprintf "%a" Asm.print_line line
+
+  let print_ins name operands = print_line (ins name operands)
+
+  let print_ins_cond name cond operands =
+    print_line (ins_cond name cond operands)
+
+  module Acc = struct
+    (* collect instructions *)
+    let asm_code = ref []
+
+    let reset_asm_code () = asm_code := []
+
+    let append line = asm_code := line :: !asm_code
+
+    let ins name operands = append (ins name operands)
+
+    let ins_cond name cond operands = append (ins_cond name cond operands)
+
+    let generate_asm oc =
+      let asm_code = List.rev !asm_code in
+      reset_asm_code ();
+      Asm.generate_asm oc asm_code
+  end
```

Comment:
(Should we use a doubly-linked list,
to avoid the need for a `rev`. I am a
bit worried we could end up with
very long lists.)

Reply:
I'm adding this task to fix later. We should do it for both targets, this is copied from somewhere in amd64 code.